File: app/src/main/java/com/elementary/tasks/core/utils/Permissions.java
/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.elementary.tasks.core.utils;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Build;

@TargetApi(Build.VERSION_CODES.M)
public class Permissions {

    public static final String READ_CONTACTS = Manifest.permission.READ_CONTACTS;
    public static final String GET_ACCOUNTS = Manifest.permission.GET_ACCOUNTS;

    public static final String READ_CALENDAR = Manifest.permission.READ_CALENDAR;
    public static final String WRITE_CALENDAR = Manifest.permission.WRITE_CALENDAR;

    public static final String WRITE_EXTERNAL = Manifest.permission.WRITE_EXTERNAL_STORAGE;
    public static final String READ_EXTERNAL = Manifest.permission.READ_EXTERNAL_STORAGE;

    public static final String ACCESS_FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION;
    public static final String ACCESS_COARSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION;

    public static final String READ_PHONE_STATE = Manifest.permission.READ_PHONE_STATE;
    public static final String CALL_PHONE = Manifest.permission.CALL_PHONE;

    public static final String SEND_SMS = Manifest.permission.SEND_SMS;
    
    public static final String MANAGE_DOCUMENTS = Manifest.permission.MANAGE_DOCUMENTS;
    public static final String READ_CALLS = Manifest.permission.READ_CALL_LOG;
    public static final String RECORD_AUDIO = Manifest.permission.RECORD_AUDIO;

    public static boolean checkPermission(Activity a, String... permissions) {
        if (!Module.isMarshmallow()) return true;

        boolean res = true;
        for (String string : permissions) {
            if (a.checkSelfPermission(string) != PackageManager.PERMISSION_GRANTED) {
                res = false;
            }
        }

        return res;
    }

    public static boolean checkPermission(Activity a, String permission) {
        return !Module.isMarshmallow() || a.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
    }

    public static void requestPermission(Activity a, int requestCode, String... permission){
        int size = permission.length;
        if (size == 1) {
            a.requestPermissions(permission, requestCode);
        } else {
            String[] array = new String[size];
            System.arraycopy(permission, 0, array, 0, size);
            a.requestPermissions(array, requestCode);
        }
    }

    public static void showInfo(Activity a, String permission){
        a.shouldShowRequestPermissionRationale(permission);
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/Permissions.kt
package com.elementary.tasks.core.utils

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat

object Permissions {

    const val READ_CONTACTS = Manifest.permission.READ_CONTACTS
    const val GET_ACCOUNTS = Manifest.permission.GET_ACCOUNTS
    const val READ_CALENDAR = Manifest.permission.READ_CALENDAR
    const val WRITE_CALENDAR = Manifest.permission.WRITE_CALENDAR
    const val WRITE_EXTERNAL = Manifest.permission.WRITE_EXTERNAL_STORAGE
    const val READ_EXTERNAL = Manifest.permission.READ_EXTERNAL_STORAGE
    const val ACCESS_FINE_LOCATION = Manifest.permission.ACCESS_FINE_LOCATION
    const val ACCESS_COARSE_LOCATION = Manifest.permission.ACCESS_COARSE_LOCATION
    @SuppressLint("InlinedApi")
    const val BACKGROUND_LOCATION = Manifest.permission.ACCESS_BACKGROUND_LOCATION
    const val READ_PHONE_STATE = Manifest.permission.READ_PHONE_STATE
    const val CALL_PHONE = Manifest.permission.CALL_PHONE
    const val RECORD_AUDIO = Manifest.permission.RECORD_AUDIO
    const val BLUETOOTH = Manifest.permission.BLUETOOTH
    const val CAMERA = Manifest.permission.CAMERA
    @RequiresApi(Build.VERSION_CODES.P)
    const val FOREGROUND = Manifest.permission.FOREGROUND_SERVICE

    fun isBgLocationAllowed(context: Context): Boolean {
        if (Module.isQ) {
            return ContextCompat.checkSelfPermission(context, BACKGROUND_LOCATION) == PackageManager.PERMISSION_GRANTED
        }
        return true
    }

    fun checkForeground(context: Context): Boolean {
        if (Module.isPie) {
            if (ContextCompat.checkSelfPermission(context, FOREGROUND) != PackageManager.PERMISSION_GRANTED) {
                return false
            }
            return true
        }
        return true
    }

    fun ensureBackgroundLocation(activity: Activity, requestCode: Int): Boolean {
        return if (isBgLocationAllowed(activity)) {
            true
        } else {
            if (Module.isQ) {
                requestPermission(activity, requestCode, BACKGROUND_LOCATION)
                false
            } else {
                return true
            }
        }
    }

    fun ensureForeground(activity: Activity, requestCode: Int): Boolean {
        return if (checkForeground(activity)) {
            true
        } else {
            if (Module.isPie) {
                requestPermission(activity, requestCode, FOREGROUND)
                false
            } else {
                return true
            }
        }
    }

    fun checkPermission(grantResults: IntArray): Boolean {
        if (grantResults.isEmpty()) {
            return false
        } else {
            for (p in grantResults) {
                if (p != PackageManager.PERMISSION_GRANTED) {
                    return false
                }
            }
            return true
        }
    }

    fun checkPermission(activity: Activity, requestCode: Int, vararg permissions: String): Boolean {
        return if (checkPermission(activity, *permissions)) {
            true
        } else {
            requestPermission(activity, requestCode, *permissions)
            false
        }
    }

    fun checkPermission(a: Context, vararg permissions: String): Boolean {
        if (!Module.isMarshmallow) {
            return true
        }
        var res = true
        for (string in permissions) {
            if (ContextCompat.checkSelfPermission(a, string) != PackageManager.PERMISSION_GRANTED) {
                res = false
            }
        }
        return res
    }

    fun checkPermission(a: Context, permission: String): Boolean {
        return !Module.isMarshmallow || ContextCompat.checkSelfPermission(a, permission) == PackageManager.PERMISSION_GRANTED
    }

    fun requestPermission(a: Activity, requestCode: Int, vararg permission: String) {
        if (Module.isMarshmallow) {
            val size = permission.size
            if (size == 1) {
                a.requestPermissions(permission, requestCode)
            } else {
                val array = arrayOfNulls<String>(size)
                System.arraycopy(permission, 0, array, 0, size)
                a.requestPermissions(array, requestCode)
            }
        }
    }
}
